### ISSUE TITLE
Crashing during Crawl

### ISSUE THREAD

**Issue by koolmutt** [johanneszab/TumblThree#57]
Not sure whats causing had no problems with Tumbl one & two. Running windows 10 64bit.
![image](https://cloud.githubusercontent.com/assets/2045923/24785879/3a8d608a-1b2b-11e7-9f7d-daeb5015c993.png) 
If I press Debug
![image](https://cloud.githubusercontent.com/assets/2045923/24785922/8b412aa2-1b2b-11e7-876d-e86cc29a03c4.png)

**Comment by johanneszab** [comment]
That happens with the v1.0.4.32?

**Comment by amigre38** [comment]
Not on my side and i have the exact same error, when by example i try to download again the blog:
01graphy.tumblr.com
few pictures inside, i have it downloaded completely, and trying with the new version (even the latest of today) i have a crash of the Tumblr.
I just try with the 1.0.4.32 and i have no crash if that can help?

**Comment by koolmutt** [comment]
Version 1.0.4.36

**Comment by johanneszab** [comment]
Thanks for reporting @koolmutt and also to @amigre38!

Please check out the newest release. I've forgotten to adjust the code to properly catch the OperationCanceledException in the async code which basically means that the application always crashed if it was stopped. I guess thats what the problem was?
Previously I've had a large big try/catch underneath all tasks (the crawler) which caught everything that wasn't explicit handled. I've re-added now at least a catch for all NullReferenceExceptions as the api data seems to be more or less reliable and I think checking every element for valid data is even worse.

The 01graphy.tumblr.com works for me now with the newest beta release (v1.0.4.39). Also if the crawler or the url grabber finishes its work.

**Comment by amigre38** [comment]
In fact the problem is not fully corrected, i have this error (i use the put in the clipboard option) because it's not somewhere in the code.

System.AggregateException was unhandled
Message: An unhandled exception of type 'System.AggregateException' occurred in mscorlib.dll
Additional information: Aucune exception de tâche n'a été observée en attendant la tâche ou en accédant à sa propriété Exception. Par conséquent, l'exception non prise en charge a été à nouveau levée par le thread finaliseur.

I have this error in this dll since few versions (maybe the ones after your big work on the code).

**Comment by johanneszab** [comment]
Its crashing the application too or it that in Visual Studio? When exactly does that happen? When the crawl is finished or you stop the crawl?

I know that i don't explicit observe the tasks exception, but for me it's handled somehow. I guess we have to check the tasks exceptions then or unwind them so that the aggregateexception is observed.

**Comment by amigre38** [comment]
I don't stop the crawling, i start it, and i let the things goes automatically.
I think it's when the blog is finishing.

I don't know how i can help you here, cause it's inside the IDE that i get the error. And after i get it i can click on continue but the code stop and i go back inside the IDE.

**Comment by johanneszab** [comment]
Ok, I'll upload something in a few minutes which probably fix this. Would be nice if you could check it!

**Comment by johanneszab** [comment]
[TumblThree-v1.0.4.40.zip](https://github.com/johanneszab/TumblThree/files/906735/TumblThree-v1.0.4.40.zip)

**Comment by amigre38** [comment]
I just download it, i add the blog to the queue, i click on crawl and i don't know why but it looks like the program is frozen.

**Comment by amigre38** [comment]
In fact it's clicking on crawl that freeze the program. Even if the queue is emty.

**Comment by johanneszab** [comment]
haha.. okay. True.

I'll figure that out I guess. Thanks though.

**Comment by amigre38** [comment]
I see the blog starting in the queue but it's like if nothing is really done, it stop at the first step (i mean i can't see it going further).

**Comment by johanneszab** [comment]
Maybe you have to fix this.

I would help, but since the .39 releases it doesn't crash for me at all. So It's quite a lot of guess work here.

Do you use new blogs or were they added before. Could you check with a newly added blog? Maybe the problem is again the path of the database..
Do you have windows 10?

If the exception if called in the microsoft libraries like above (_An unhandled exception of type 'System.AggregateException' occurred in mscorlib.dll_), thats most likely because the exception is not handled and it's particular hard to guess if the code is _async_, because the debugger then points to completely different locations as of where the exception actually occurred.
It might help if you just put everywhere a large try/catch around the whole async method or wherever tasks are started with Task.Run().

**Comment by johanneszab** [comment]
Ah, the easiest thing probably is, if its still the AggregateException, is to look in the variables window which is usually in the bottom left.
There click on the Exception, then look into the innerException. That should help to debug.

**Comment by amigre38** [comment]
In fact i don't have crash anymore, it's blocked on this without moving:
![image](https://cloud.githubusercontent.com/assets/26059690/24837368/06ef5bca-1d33-11e7-912d-83025ec8c027.png)

I'm on windows 10 and i will try to test using the source inside the Visual Studio IDE

**Comment by johanneszab** [comment]
So, I don't know if this is fixed or still there..? I'll close it since I probably won't fix it without more information, since it simply does not appear here.

If it's still there, please re-open with more information.
